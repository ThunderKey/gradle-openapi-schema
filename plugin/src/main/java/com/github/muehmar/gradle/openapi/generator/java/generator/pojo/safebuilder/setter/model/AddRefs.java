package com.github.muehmar.gradle.openapi.generator.java.generator.pojo.safebuilder.setter.model;

import io.github.muehmar.codegenerator.writer.Writer;

public interface AddRefs {
  Writer addRefs(Writer writer);
}
