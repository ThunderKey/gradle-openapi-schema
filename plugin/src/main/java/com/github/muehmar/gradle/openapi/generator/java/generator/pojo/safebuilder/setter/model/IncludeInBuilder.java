package com.github.muehmar.gradle.openapi.generator.java.generator.pojo.safebuilder.setter.model;

public interface IncludeInBuilder<T extends SetterMember> {
  boolean includeInBuilder(T member);
}
