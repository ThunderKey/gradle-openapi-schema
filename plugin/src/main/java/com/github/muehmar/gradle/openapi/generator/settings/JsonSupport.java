package com.github.muehmar.gradle.openapi.generator.settings;

import ch.bluecare.commons.data.PList;
import java.io.Serializable;
import java.util.Optional;

public enum JsonSupport implements Serializable {
  JACKSON("jackson"),
  NONE("none");

  private final String value;

  JsonSupport(String value) {
    this.value = value;
  }

  public String getValue() {
    return value;
  }

  public static Optional<JsonSupport> fromString(String value) {
    return PList.of(values()).find(support -> support.value.equalsIgnoreCase(value));
  }
}
