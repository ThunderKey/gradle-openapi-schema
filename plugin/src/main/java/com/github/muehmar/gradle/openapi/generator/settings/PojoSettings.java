package com.github.muehmar.gradle.openapi.generator.settings;

import ch.bluecare.commons.data.PList;
import io.github.muehmar.pojoextension.annotations.PojoExtension;
import java.io.Serializable;
import java.util.List;
import java.util.Objects;

@PojoExtension
public class PojoSettings extends PojoSettingsExtension implements Serializable {
  private final JsonSupport jsonSupport;
  private final String packageName;
  private final String suffix;
  private final boolean enableSafeBuilder;
  private final boolean enableConstraints;
  private final List<ClassTypeMapping> classTypeMappings;
  private final List<FormatTypeMapping> formatTypeMappings;
  private final EnumDescriptionSettings enumDescriptionSettings;

  PojoSettings(
      JsonSupport jsonSupport,
      String packageName,
      String suffix,
      boolean enableSafeBuilder,
      boolean enableConstraints,
      List<ClassTypeMapping> classTypeMappings,
      List<FormatTypeMapping> formatTypeMappings,
      EnumDescriptionSettings enumDescriptionSettings) {
    this.jsonSupport = jsonSupport;
    this.packageName = packageName;
    this.suffix = suffix;
    this.enableSafeBuilder = enableSafeBuilder;
    this.enableConstraints = enableConstraints;
    this.classTypeMappings = classTypeMappings;
    this.formatTypeMappings = formatTypeMappings;
    this.enumDescriptionSettings = enumDescriptionSettings;
  }

  public PojoSettings(
      JsonSupport jsonSupport,
      String packageName,
      String suffix,
      boolean enableSafeBuilder,
      boolean enableConstraints,
      PList<ClassTypeMapping> classTypeMappings,
      PList<FormatTypeMapping> formatTypeMappings,
      EnumDescriptionSettings enumDescriptionSettings) {
    this.jsonSupport = jsonSupport;
    this.packageName = packageName;
    this.suffix = suffix;
    this.enableSafeBuilder = enableSafeBuilder;
    this.enableConstraints = enableConstraints;
    this.classTypeMappings = classTypeMappings.toArrayList();
    this.formatTypeMappings = formatTypeMappings.toArrayList();
    this.enumDescriptionSettings = enumDescriptionSettings;
  }

  public String getPackageName() {
    return packageName;
  }

  public String getSuffix() {
    return suffix;
  }

  public JsonSupport getJsonSupport() {
    return jsonSupport;
  }

  public boolean isJacksonJson() {
    return jsonSupport.equals(JsonSupport.JACKSON);
  }

  public boolean isEnableConstraints() {
    return enableConstraints;
  }

  public PList<ClassTypeMapping> getClassTypeMappings() {
    return PList.fromIter(classTypeMappings);
  }

  public PList<FormatTypeMapping> getFormatTypeMappings() {
    return PList.fromIter(formatTypeMappings);
  }

  public boolean isEnableSafeBuilder() {
    return enableSafeBuilder;
  }

  public boolean isDisableSafeBuilder() {
    return !isEnableSafeBuilder();
  }

  public EnumDescriptionSettings getEnumDescriptionSettings() {
    return enumDescriptionSettings;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    PojoSettings that = (PojoSettings) o;
    return enableSafeBuilder == that.enableSafeBuilder
        && enableConstraints == that.enableConstraints
        && jsonSupport == that.jsonSupport
        && Objects.equals(packageName, that.packageName)
        && Objects.equals(suffix, that.suffix)
        && Objects.equals(classTypeMappings, that.classTypeMappings)
        && Objects.equals(formatTypeMappings, that.formatTypeMappings)
        && Objects.equals(enumDescriptionSettings, that.enumDescriptionSettings);
  }

  @Override
  public int hashCode() {
    return Objects.hash(
        jsonSupport,
        packageName,
        suffix,
        enableSafeBuilder,
        enableConstraints,
        classTypeMappings,
        formatTypeMappings,
        enumDescriptionSettings);
  }

  @Override
  public String toString() {
    return "PojoSettings{"
        + "jsonSupport="
        + jsonSupport
        + ", packageName='"
        + packageName
        + '\''
        + ", suffix='"
        + suffix
        + '\''
        + ", enableSafeBuilder="
        + enableSafeBuilder
        + ", enableConstraints="
        + enableConstraints
        + ", classTypeMappings="
        + classTypeMappings
        + ", formatTypeMappings="
        + formatTypeMappings
        + ", enumDescriptionExtraction="
        + enumDescriptionSettings
        + '}';
  }
}
