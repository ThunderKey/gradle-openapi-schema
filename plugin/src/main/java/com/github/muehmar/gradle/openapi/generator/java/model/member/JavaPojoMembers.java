package com.github.muehmar.gradle.openapi.generator.java.model.member;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaObjectPojo;
import com.github.muehmar.gradle.openapi.generator.model.Necessity;
import com.github.muehmar.gradle.openapi.generator.model.Nullability;
import java.util.function.UnaryOperator;
import lombok.EqualsAndHashCode;
import lombok.ToString;

/**
 * The members of a {@link JavaObjectPojo}. This container ensures, that the same property does not
 * exist twice. Adding a property which is already present will not add a new property/member, it
 * will keep the current property but maybe with lower restrictions concerning {@link Nullability}
 * and {@link Necessity}.
 */
@EqualsAndHashCode
@ToString
public class JavaPojoMembers {
  private final PList<JavaPojoMember> members;

  private JavaPojoMembers(PList<JavaPojoMember> members) {
    this.members = members;
  }

  public static JavaPojoMembers empty() {
    return new JavaPojoMembers(PList.empty());
  }

  public static JavaPojoMembers fromMembers(PList<JavaPojoMember> members) {
    return JavaPojoMembers.empty().add(members);
  }

  public JavaPojoMembers add(JavaPojoMember memberToAdd) {
    final boolean hasMemberWithSameKey =
        members.exists(member -> member.getMemberKey().equals(memberToAdd.getMemberKey()));
    if (hasMemberWithSameKey) {
      final PList<JavaPojoMember> newMembers =
          members.map(member -> member.mergeToLeastRestrictive(memberToAdd).orElse(member));
      return new JavaPojoMembers(newMembers);
    } else {
      return new JavaPojoMembers(members.add(memberToAdd));
    }
  }

  public JavaPojoMembers add(PList<JavaPojoMember> membersToAdd) {
    return membersToAdd.foldLeft(this, JavaPojoMembers::add);
  }

  public JavaPojoMembers add(JavaPojoMembers membersToAdd) {
    return add(membersToAdd.asList());
  }

  public JavaPojoMembers map(UnaryOperator<JavaPojoMember> f) {
    return new JavaPojoMembers(members.map(f));
  }

  public PList<JavaPojoMember> asList() {
    return members;
  }

  public boolean isEmpty() {
    return members.isEmpty();
  }

  public boolean hasRequiredMembers() {
    return members.exists(JavaPojoMember::isRequired);
  }

  public int getRequiredMemberCount() {
    return members.filter(JavaPojoMember::isRequired).size();
  }
}
