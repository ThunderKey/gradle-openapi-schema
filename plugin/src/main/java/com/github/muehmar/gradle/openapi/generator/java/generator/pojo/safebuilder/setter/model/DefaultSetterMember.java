package com.github.muehmar.gradle.openapi.generator.java.generator.pojo.safebuilder.setter.model;

import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;

public abstract class DefaultSetterMember implements SetterMember {
  @Override
  public String builderMethodName(PojoSettings settings) {
    return getMember().prefixedMethodName(settings.getBuilderMethodPrefix()).asString();
  }

  @Override
  public String argumentType() {
    return getMember().getJavaType().getParameterizedClassName().asString();
  }
}
