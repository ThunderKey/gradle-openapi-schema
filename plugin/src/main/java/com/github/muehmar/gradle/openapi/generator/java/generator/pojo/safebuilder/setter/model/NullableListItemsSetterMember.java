package com.github.muehmar.gradle.openapi.generator.java.generator.pojo.safebuilder.setter.model;

import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;

public abstract class NullableListItemsSetterMember implements SetterMember {

  @Override
  public String builderMethodName(PojoSettings settings) {
    return getMember().prefixedMethodName(settings.getBuilderMethodPrefix()).asString().concat("_");
  }

  @Override
  public String argumentType() {
    return getMember()
        .getJavaType()
        .getParameterizedClassName()
        .asStringWrappingNullableValueType();
  }
}
