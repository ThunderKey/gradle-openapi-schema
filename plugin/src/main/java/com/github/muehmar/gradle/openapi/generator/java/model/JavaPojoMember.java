package com.github.muehmar.gradle.openapi.generator.java.model;

import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaType;
import com.github.muehmar.gradle.openapi.generator.model.Name;
import com.github.muehmar.gradle.openapi.generator.model.Necessity;
import com.github.muehmar.gradle.openapi.generator.model.Nullability;
import com.github.muehmar.gradle.openapi.generator.model.PojoMember;
import com.github.muehmar.gradle.openapi.generator.settings.GetterSuffixes;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@EqualsAndHashCode
@ToString
public class JavaPojoMember {
  private final Name name;
  private final String description;
  private final JavaType javaType;
  private final Necessity necessity;
  private final Nullability nullability;

  private JavaPojoMember(
      Name name,
      String description,
      JavaType javaType,
      Necessity necessity,
      Nullability nullability) {
    this.javaType = javaType;
    this.name = name;
    this.description = description;
    this.necessity = necessity;
    this.nullability = nullability;
  }

  public static JavaPojoMember of(
      Name name,
      String description,
      JavaType javaType,
      Necessity necessity,
      Nullability nullability) {
    return new JavaPojoMember(name, description, javaType, necessity, nullability);
  }

  public static JavaPojoMember wrap(PojoMember pojoMember, TypeMappings typeMappings) {
    final JavaType javaType = JavaType.wrap(pojoMember.getType(), typeMappings);
    return new JavaPojoMember(
        pojoMember.getName(),
        pojoMember.getDescription(),
        javaType,
        pojoMember.getNecessity(),
        pojoMember.getNullability());
  }

  public Nullability getNullability() {
    return nullability;
  }

  public Necessity getNecessity() {
    return necessity;
  }

  public String getDescription() {
    return description;
  }

  public Name getName() {
    return name;
  }

  public JavaType getJavaType() {
    return javaType;
  }

  public boolean isOptional() {
    return necessity.isOptional();
  }

  public boolean isRequired() {
    return necessity.isRequired();
  }

  public boolean isNullable() {
    return nullability.isNullable();
  }

  public boolean isNotNullable() {
    return nullability.isNotNullable();
  }

  public boolean isRequiredAndNullable() {
    return isRequired() && isNullable();
  }

  public boolean isRequiredAndNotNullable() {
    return isRequired() && isNotNullable();
  }

  public boolean isOptionalAndNullable() {
    return isOptional() && isNullable();
  }

  public boolean isOptionalAndNotNullable() {
    return isOptional() && isNotNullable();
  }

  public Name getGetterName() {
    return prefixedMethodName("get");
  }

  public Name getGetterNameWithSuffix(PojoSettings settings) {
    return getGetterName().append(determineSuffix(settings));
  }

  private String determineSuffix(PojoSettings settings) {
    final GetterSuffixes getterSuffixes = settings.getGetterSuffixes();
    if (isRequiredAndNotNullable()) {
      return getterSuffixes.getRequiredSuffix();
    } else if (isRequiredAndNullable()) {
      return getterSuffixes.getRequiredNullableSuffix();
    } else if (isOptionalAndNotNullable()) {
      return getterSuffixes.getOptionalSuffix();
    } else {
      return getterSuffixes.getOptionalNullableSuffix();
    }
  }

  public Name getWitherName() {
    return prefixedMethodName("with");
  }

  public Name prefixedMethodName(String prefix) {
    if (prefix.isEmpty()) {
      return name;
    } else {
      return name.startUpperCase().prefix(prefix);
    }
  }
}
