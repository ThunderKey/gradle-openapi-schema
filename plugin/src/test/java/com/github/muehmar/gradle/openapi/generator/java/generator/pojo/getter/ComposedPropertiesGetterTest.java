package com.github.muehmar.gradle.openapi.generator.java.generator.pojo.getter;

import static com.github.muehmar.gradle.openapi.SnapshotUtil.writerSnapshot;
import static com.github.muehmar.gradle.openapi.generator.settings.TestPojoSettings.defaultTestSettings;
import static io.github.muehmar.codegenerator.writer.Writer.javaWriter;
import static org.junit.jupiter.api.Assertions.assertEquals;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMember;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMembers;
import com.github.muehmar.gradle.openapi.generator.settings.JsonSupport;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.writer.Writer;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

@ExtendWith(SnapshotExtension.class)
class ComposedPropertiesGetterTest {
  private Expect expect;

  @Test
  void generate_when_objectMember_then_noOutput() {
    final Generator<JavaPojoMember, PojoSettings> generator =
        ComposedPropertiesGetter.composedPropertiesGetterGenerator();

    final Writer writer =
        generator.generate(
            JavaPojoMembers.requiredBirthdate(), defaultTestSettings(), javaWriter());

    assertEquals("", writer.asString());
  }

  @Test
  void generate_when_allOfMember_then_noOutput() {
    final Generator<JavaPojoMember, PojoSettings> generator =
        ComposedPropertiesGetter.composedPropertiesGetterGenerator();

    final Writer writer =
        generator.generate(
            JavaPojoMembers.requiredBirthdate().asAllOfMember(),
            defaultTestSettings(),
            javaWriter());

    assertEquals("", writer.asString());
  }

  @Test
  void generate_when_anyOfMemberAndNoJsonSupport_then_noOutput() {
    final Generator<JavaPojoMember, PojoSettings> generator =
        ComposedPropertiesGetter.composedPropertiesGetterGenerator();

    final Writer writer =
        generator.generate(
            JavaPojoMembers.requiredBirthdate().asAnyOfMember(),
            defaultTestSettings().withJsonSupport(JsonSupport.NONE),
            javaWriter());

    assertEquals("", writer.asString());
  }

  @Test
  @SnapshotName("requiredOneOfMember")
  void generate_when_requiredOneOfMember_then_correctOutput() {
    final Generator<JavaPojoMember, PojoSettings> generator =
        ComposedPropertiesGetter.composedPropertiesGetterGenerator();

    final Writer writer =
        generator.generate(
            JavaPojoMembers.requiredBirthdate().asOneOfMember(),
            defaultTestSettings(),
            javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("requiredAnyOfMember")
  void generate_when_requiredAnyOfMember_then_correctOutput() {
    final Generator<JavaPojoMember, PojoSettings> generator =
        ComposedPropertiesGetter.composedPropertiesGetterGenerator();

    final Writer writer =
        generator.generate(
            JavaPojoMembers.requiredBirthdate().asAnyOfMember(),
            defaultTestSettings(),
            javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("requiredNullableOneOfMember")
  void generate_when_requiredNullableOneOfMember_then_correctOutput() {
    final Generator<JavaPojoMember, PojoSettings> generator =
        ComposedPropertiesGetter.composedPropertiesGetterGenerator();

    final Writer writer =
        generator.generate(
            JavaPojoMembers.requiredNullableBirthdate().asOneOfMember(),
            defaultTestSettings(),
            javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("requiredNullableAnyOfMember")
  void generate_when_requiredNullableAnyOfMember_then_correctOutput() {
    final Generator<JavaPojoMember, PojoSettings> generator =
        ComposedPropertiesGetter.composedPropertiesGetterGenerator();

    final Writer writer =
        generator.generate(
            JavaPojoMembers.requiredNullableBirthdate().asAnyOfMember(),
            defaultTestSettings(),
            javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("optionalOneOfMember")
  void generate_when_optionalOneOfMember_then_correctOutput() {
    final Generator<JavaPojoMember, PojoSettings> generator =
        ComposedPropertiesGetter.composedPropertiesGetterGenerator();

    final Writer writer =
        generator.generate(
            JavaPojoMembers.optionalBirthdate().asOneOfMember(),
            defaultTestSettings(),
            javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("optionalAnyOfMember")
  void generate_when_optionalAnyOfMember_then_correctOutput() {
    final Generator<JavaPojoMember, PojoSettings> generator =
        ComposedPropertiesGetter.composedPropertiesGetterGenerator();

    final Writer writer =
        generator.generate(
            JavaPojoMembers.optionalBirthdate().asOneOfMember(),
            defaultTestSettings(),
            javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("optionalNullableOneOfMember")
  void generate_when_optionalNullableOneOfMember_then_correctOutput() {
    final Generator<JavaPojoMember, PojoSettings> generator =
        ComposedPropertiesGetter.composedPropertiesGetterGenerator();

    final Writer writer =
        generator.generate(
            JavaPojoMembers.optionalNullableBirthdate().asOneOfMember(),
            defaultTestSettings(),
            javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("optionalNullableAnyOfMember")
  void generate_when_optionalNullableAnyOfMember_then_correctOutput() {
    final Generator<JavaPojoMember, PojoSettings> generator =
        ComposedPropertiesGetter.composedPropertiesGetterGenerator();

    final Writer writer =
        generator.generate(
            JavaPojoMembers.optionalNullableBirthdate().asOneOfMember(),
            defaultTestSettings(),
            javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }
}
