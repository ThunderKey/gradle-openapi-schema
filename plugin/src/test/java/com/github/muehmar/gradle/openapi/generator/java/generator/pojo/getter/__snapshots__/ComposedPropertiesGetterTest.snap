anyOfMemberAndNoJsonSupport=[
java.time.LocalDate

LocalDate getBirthdateJson() {
  return birthdate;
}

LocalDate getBirthdate() {
  return birthdate;
}
]


optionalAnyOfMember=[
com.fasterxml.jackson.annotation.JsonIgnore
com.fasterxml.jackson.annotation.JsonInclude
com.fasterxml.jackson.annotation.JsonProperty
java.time.LocalDate
java.util.Optional

@JsonProperty("birthdate")
@JsonInclude(JsonInclude.Include.NON_NULL)
LocalDate getBirthdateJson() {
  return birthdate;
}

@JsonIgnore
Optional<LocalDate> getBirthdate() {
  return Optional.ofNullable(birthdate);
}
]


optionalNullableAnyOfMember=[
com.fasterxml.jackson.annotation.JsonIgnore
com.fasterxml.jackson.annotation.JsonInclude
com.fasterxml.jackson.annotation.JsonProperty
com.github.muehmar.openapi.util.JacksonNullContainer
com.github.muehmar.openapi.util.Tristate
java.time.LocalDate

@JsonProperty("birthdate")
@JsonInclude(JsonInclude.Include.NON_NULL)
Object getBirthdateJson() {
  return isBirthdateNull ? new JacksonNullContainer<>(birthdate) : birthdate;
}

@JsonIgnore
Tristate<LocalDate> getBirthdate() {
  return Tristate.ofNullableAndNullFlag(birthdate, isBirthdateNull);
}
]


optionalNullableOneOfMember=[
com.fasterxml.jackson.annotation.JsonIgnore
com.fasterxml.jackson.annotation.JsonInclude
com.fasterxml.jackson.annotation.JsonProperty
com.github.muehmar.openapi.util.JacksonNullContainer
com.github.muehmar.openapi.util.Tristate
java.time.LocalDate

@JsonProperty("birthdate")
@JsonInclude(JsonInclude.Include.NON_NULL)
Object getBirthdateJson() {
  return isBirthdateNull ? new JacksonNullContainer<>(birthdate) : birthdate;
}

@JsonIgnore
Tristate<LocalDate> getBirthdate() {
  return Tristate.ofNullableAndNullFlag(birthdate, isBirthdateNull);
}
]


optionalOneOfMember=[
com.fasterxml.jackson.annotation.JsonIgnore
com.fasterxml.jackson.annotation.JsonInclude
com.fasterxml.jackson.annotation.JsonProperty
java.time.LocalDate
java.util.Optional

@JsonProperty("birthdate")
@JsonInclude(JsonInclude.Include.NON_NULL)
LocalDate getBirthdateJson() {
  return birthdate;
}

@JsonIgnore
Optional<LocalDate> getBirthdate() {
  return Optional.ofNullable(birthdate);
}
]


requiredAnyOfMember=[
com.fasterxml.jackson.annotation.JsonIgnore
com.fasterxml.jackson.annotation.JsonInclude
com.fasterxml.jackson.annotation.JsonProperty
java.time.LocalDate

@JsonProperty("birthdate")
@JsonInclude(JsonInclude.Include.NON_NULL)
LocalDate getBirthdateJson() {
  return birthdate;
}

@JsonIgnore
LocalDate getBirthdate() {
  return birthdate;
}
]


requiredNullableAnyOfMember=[
com.fasterxml.jackson.annotation.JsonIgnore
com.fasterxml.jackson.annotation.JsonInclude
com.fasterxml.jackson.annotation.JsonProperty
com.github.muehmar.openapi.util.JacksonNullContainer
java.time.LocalDate
java.util.Optional

@JsonProperty("birthdate")
@JsonInclude(JsonInclude.Include.NON_NULL)
Object getBirthdateJson() {
  return isBirthdatePresent ? new JacksonNullContainer<>(birthdate) : null;
}

@JsonIgnore
Optional<LocalDate> getBirthdate() {
  return Optional.ofNullable(birthdate);
}
]


requiredNullableOneOfMember=[
com.fasterxml.jackson.annotation.JsonIgnore
com.fasterxml.jackson.annotation.JsonInclude
com.fasterxml.jackson.annotation.JsonProperty
com.github.muehmar.openapi.util.JacksonNullContainer
java.time.LocalDate
java.util.Optional

@JsonProperty("birthdate")
@JsonInclude(JsonInclude.Include.NON_NULL)
Object getBirthdateJson() {
  return isBirthdatePresent ? new JacksonNullContainer<>(birthdate) : null;
}

@JsonIgnore
Optional<LocalDate> getBirthdate() {
  return Optional.ofNullable(birthdate);
}
]


requiredOneOfMember=[
com.fasterxml.jackson.annotation.JsonIgnore
com.fasterxml.jackson.annotation.JsonInclude
com.fasterxml.jackson.annotation.JsonProperty
java.time.LocalDate

@JsonProperty("birthdate")
@JsonInclude(JsonInclude.Include.NON_NULL)
LocalDate getBirthdateJson() {
  return birthdate;
}

@JsonIgnore
LocalDate getBirthdate() {
  return birthdate;
}
]